guides/keras_nlp/fnet_classification.py

```diff
@@ -0,0 +1,380 @@
+
+Note: To learn more about the Transformer architecture, please visit [Jay Alammar's
+peerless blog](https://jalammar.github.io/illustrated-transformer/). 
+
+Recently, there has been an effort to reduce the time complexity of the self-attention
+mechanism and improve performance without sacrificing the quality of results.
+Models such as Longformers, Reformers, etc. come to mind. In 2020, a paper titled
+[FNet: Mixing Tokens with Fourier Transforms](https://arxiv.org/abs/2105.03824)
+replaced the self-attention layer in BERT with a Fourier Transform layer. The
+magnitude of this was not lost on the research community; replacing a relatively
+complicated self-attention layer with a simpler "token mixing" layer resulted in
+comparable accuracy and a speed-up during training. This opens up further avenues
+of research - can we replace the self-attention layer with any simple "token mixing"
+layer and get comparable results?
+
+A couple of points from the paper stood out:
+1. The authors claim that FNet is 80% faster than BERT on GPUs and 70% faster on TPUs.
+The reason for this speed-up is two-fold:
+    a. The Fourier Transform layer is unparametrized, it does not have any parameters!
+    b. The authors use Fast Fourier Transform (FFT); this reduces the time complexity
+    from O(n^2) (in the case of self-attention) to O(n log n).
+2. What's astounding is that FNet still manages to achieve 92-97% of the accuracy
+of BERT on the GLUE benchmark.
+"""
+
+"""
+## Setup
+
+Before we start with the implementation, let's install the KerasNLP library, and
+import all the necessary packages.
+"""
+
+"""shell
+pip install -q keras-nlp
+pip install -q tfds-nightly
+"""
+import keras_nlp
+import numpy as np
+import pathlib
+import random
+import tensorflow as tf
+import tensorflow_datasets as tfds
+
+from tensorflow import keras
+from tensorflow_text.tools.wordpiece_vocab import bert_vocab_from_dataset as bert_vocab
+
+"""
+Let's also define our parameters/hyperparameters.
+"""
+BATCH_SIZE = 64
+EPOCHS = 3
+MAX_SEQUENCE_LENGTH = 40
+VOCAB_SIZE = 15000
+
+EMBED_DIM = 128
+INTERMEDIATE_DIM = 512
+
+"""
+## Loading the Dataset
+Tensorflow Datasets (TFDS) is a library that provides a unified API for working
+with datasets that are stored in the TensorFlow format. We will use TFDS to load
+the SST-2 dataset.
+"""
+train_ds, val_ds, test_ds = tfds.load(
+    "huggingface:sst", split=["train", "validation", "test"], shuffle_files=True
+)
+
+"""
+## Preprocessing the Dataset
+We need to perform these two steps:
+1. Convert the input sentences to lowercase.
+2. The label is a sentiment score (float) lying in the range [0, 1]. We will
+convert it to either 0 or 1, keeping 0.5 as the threshold. This is a necessary
+step because the task we want to solve is a classification task.
+"""
+
+
+def preprocess_dataset(dataset):
+    dataset = dataset.map(
+        lambda x: {
+            "sentence": tf.strings.lower(x["sentence"]),
```

Comment:
Can't we lowercase inside the tokenizer? Why do that here?

Reply:
It lowercases the special tokens as well, which I wanted to avoid.